File: tui/src/api/kubernetes.zig
const std = @import("std");
const config = @import("../config.zig");
const http = @import("http_client.zig");
const Allocator = std.mem.Allocator;

pub const K8sBackupEntry = struct {
    name: []const u8,
    namespace: []const u8,
    source_type: []const u8, // "VolSync" or "Velero"
    status: []const u8,
    schedule: []const u8,
    last_run: []const u8,
};

pub const DetectedProviders = struct {
    volsync: bool = false,
    velero: bool = false,
};

pub const KubeClient = struct {
    allocator: Allocator,
    kubeconfig: []const u8,

    pub fn init(allocator: Allocator, kubeconfig: []const u8) KubeClient {
        return .{
            .allocator = allocator,
            .kubeconfig = kubeconfig,
        };
    }

    /// Detect which backup providers (VolSync, Velero) are installed by checking CRDs.
    pub fn detectProviders(self: *KubeClient) DetectedProviders {
        const output = self.runKubectl(&.{
            "get", "crd", "--no-headers", "-o", "custom-columns=NAME:.metadata.name",
        }) orelse return .{};
        defer self.allocator.free(output);

        var result = DetectedProviders{};
        var lines = std.mem.splitScalar(u8, output, '\n');
        while (lines.next()) |line| {
            const trimmed = std.mem.trim(u8, line, " \t\r");
            if (trimmed.len == 0) continue;
            if (std.mem.indexOf(u8, trimmed, "volsync") != null) result.volsync = true;
            if (std.mem.indexOf(u8, trimmed, "velero") != null) result.velero = true;
        }
        return result;
    }

    /// Fetch VolSync ReplicationSources across all namespaces.
    pub fn getVolsyncSources(self: *KubeClient) []K8sBackupEntry {
        const output = self.runKubectl(&.{
            "get", "replicationsources.volsync.backube", "-A", "-o", "json",
        }) orelse return &.{};
        defer self.allocator.free(output);

        return self.parseK8sBackups(output, "VolSync");
    }

    /// Fetch Velero Backups across all namespaces.
    pub fn getVeleroBackups(self: *KubeClient) []K8sBackupEntry {
        const output = self.runKubectl(&.{
            "get", "backups.velero.io", "-A", "-o", "json",
        }) orelse return &.{};
        defer self.allocator.free(output);

        return self.parseK8sBackups(output, "Velero");
    }

    fn parseK8sBackups(self: *KubeClient, output: []const u8, source_type: []const u8) []K8sBackupEntry {
        var parsed = std.json.parseFromSlice(std.json.Value, self.allocator, output, .{
            .ignore_unknown_fields = true,
            .allocate = .alloc_always,
        }) catch return &.{};
        defer parsed.deinit();

        const root = switch (parsed.value) {
            .object => |obj| obj,
            else => return &.{},
        };

        const items = switch (root.get("items") orelse return &.{}) {
            .array => |arr| arr.items,
            else => return &.{},
        };

        var results: std.ArrayListUnmanaged(K8sBackupEntry) = .empty;
        for (items) |item| {
            const obj = switch (item) {
                .object => |o| o,
                else => continue,
            };

            const metadata = switch (obj.get("metadata") orelse continue) {
                .object => |o| o,
                else => continue,
            };

            const name = self.allocator.dupe(u8, http.jsonStr(metadata, "name", "unknown")) catch continue;
            const namespace = self.allocator.dupe(u8, http.jsonStr(metadata, "namespace", "default")) catch continue;

            // Extract status and schedule based on source type
            var status: []const u8 = undefined;
            var schedule: []const u8 = undefined;
            var last_run: []const u8 = undefined;

            if (std.mem.eql(u8, source_type, "VolSync")) {
                status = self.parseVolsyncStatus(obj);
                schedule = self.parseVolsyncSchedule(obj);
                last_run = self.parseVolsyncLastRun(obj);
            } else {
                status = self.parseVeleroStatus(obj);
                schedule = self.parseVeleroSchedule(obj);
                last_run = self.parseVeleroLastRun(obj);
            }

            results.append(self.allocator, .{
                .name = name,
                .namespace = namespace,
                .source_type = self.allocator.dupe(u8, source_type) catch continue,
                .status = status,
                .schedule = schedule,
                .last_run = last_run,
            }) catch continue;
        }

        return results.toOwnedSlice(self.allocator) catch &.{};
    }

    fn parseVolsyncStatus(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const status_obj = switch (obj.get("status") orelse return self.allocator.dupe(u8, "unknown") catch "unknown") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "unknown") catch "unknown",
        };

        // Check conditions array for Synchronizing condition
        const conditions = switch (status_obj.get("conditions") orelse return self.allocator.dupe(u8, "unknown") catch "unknown") {
            .array => |arr| arr.items,
            else => return self.allocator.dupe(u8, "unknown") catch "unknown",
        };

        for (conditions) |cond| {
            const cond_obj = switch (cond) {
                .object => |o| o,
                else => continue,
            };
            const cond_type = http.jsonStr(cond_obj, "type", "");
            if (std.mem.eql(u8, cond_type, "Synchronizing")) {
                const cond_status = http.jsonStr(cond_obj, "status", "Unknown");
                if (std.mem.eql(u8, cond_status, "True")) {
                    return self.allocator.dupe(u8, "Syncing") catch "Syncing";
                }
                return self.allocator.dupe(u8, "Idle") catch "Idle";
            }
        }

        return self.allocator.dupe(u8, "unknown") catch "unknown";
    }

    fn parseVolsyncSchedule(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const spec = switch (obj.get("spec") orelse return self.allocator.dupe(u8, "-") catch "-") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "-") catch "-",
        };
        const trigger = switch (spec.get("trigger") orelse return self.allocator.dupe(u8, "-") catch "-") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "-") catch "-",
        };
        return self.allocator.dupe(u8, http.jsonStr(trigger, "schedule", "-")) catch "-";
    }

    fn parseVolsyncLastRun(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const status_obj = switch (obj.get("status") orelse return self.allocator.dupe(u8, "-") catch "-") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "-") catch "-",
        };
        return self.allocator.dupe(u8, http.jsonStr(status_obj, "lastSyncTime", "-")) catch "-";
    }

    fn parseVeleroStatus(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const status_obj = switch (obj.get("status") orelse return self.allocator.dupe(u8, "unknown") catch "unknown") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "unknown") catch "unknown",
        };
        return self.allocator.dupe(u8, http.jsonStr(status_obj, "phase", "unknown")) catch "unknown";
    }

    fn parseVeleroSchedule(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const spec = switch (obj.get("spec") orelse return self.allocator.dupe(u8, "-") catch "-") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "-") catch "-",
        };
        return self.allocator.dupe(u8, http.jsonStr(spec, "scheduleName", "-")) catch "-";
    }

    fn parseVeleroLastRun(self: *KubeClient, obj: std.json.ObjectMap) []const u8 {
        const status_obj = switch (obj.get("status") orelse return self.allocator.dupe(u8, "-") catch "-") {
            .object => |o| o,
            else => return self.allocator.dupe(u8, "-") catch "-",
        };
        return self.allocator.dupe(u8, http.jsonStr(status_obj, "completionTimestamp", "-")) catch "-";
    }

    /// Run a kubectl command with standard flags and return stdout.
    /// Returns null on any failure.
    fn runKubectl(self: *KubeClient, extra_args: []const []const u8) ?[]const u8 {
        var argv: std.ArrayListUnmanaged([]const u8) = .empty;
        defer argv.deinit(self.allocator);

        argv.append(self.allocator, "kubectl") catch return null;
        argv.appendSlice(self.allocator, extra_args) catch return null;
        argv.appendSlice(self.allocator, &.{
            "--kubeconfig", self.kubeconfig,
        }) catch return null;

        const result = std.process.Child.run(.{
            .allocator = self.allocator,
            .argv = argv.items,
            .max_output_bytes = 512 * 1024,
        }) catch return null;
        defer self.allocator.free(result.stderr);

        const term = result.term;
        if (term == .Exited and term.Exited == 0) {
            return result.stdout;
        }

        self.allocator.free(result.stdout);
        return null;
    }

    pub fn deinit(self: *KubeClient) void {
        _ = self;
    }
};

/// Derive the kubeconfig path from a talos config path.
/// Given "~/talos/apollo/talosconfig", returns "~/talos/apollo/kubeconfig".
pub fn deriveKubeconfig(allocator: Allocator, talos_config_path: []const u8) ?[]const u8 {
    // Find the last path separator
    const dir_end = std.mem.lastIndexOfScalar(u8, talos_config_path, '/') orelse return null;
    return std.fmt.allocPrint(allocator, "{s}/kubeconfig", .{talos_config_path[0..dir_end]}) catch null;
}

test "deriveKubeconfig" {
    const alloc = std.testing.allocator;
    const result = deriveKubeconfig(alloc, "~/talos/apollo/talosconfig") orelse unreachable;
    defer alloc.free(result);
    try std.testing.expectEqualStrings("~/talos/apollo/kubeconfig", result);
}

File: tui/src/integrations/kubernetes.rs
use crate::integrations::command;
use anyhow::Result;
use serde_json::Value;
use std::{env, path::PathBuf};

#[derive(Debug, Clone, Default)]
pub struct K8sBackupEntry {
    pub name: String,
    pub namespace: String,
    pub source_type: String,
    pub status: String,
    pub schedule: String,
    pub last_run: String,
}

#[derive(Debug, Clone, Default)]
pub struct DetectedProviders {
    pub volsync: bool,
    pub velero: bool,
}

#[derive(Debug, Clone, Default)]
pub struct ClusterNode {
    pub name: String,
    pub internal_ip: Option<String>,
    pub role: String,
}

pub fn detect_providers() -> Result<DetectedProviders> {
    let output = kubectl(&[
        "get",
        "crd",
        "--no-headers",
        "-o",
        "custom-columns=NAME:.metadata.name",
    ])?;
    let mut providers = DetectedProviders::default();
    for line in output.lines() {
        let line = line.trim();
        if line.contains("volsync") {
            providers.volsync = true;
        }
        if line.contains("velero") {
            providers.velero = true;
        }
    }
    Ok(providers)
}

pub fn get_volsync_sources() -> Result<Vec<K8sBackupEntry>> {
    parse_backups(
        &kubectl(&[
            "get",
            "replicationsources.volsync.backube",
            "-A",
            "-o",
            "json",
        ])?,
        "VolSync",
    )
}

pub fn get_velero_backups() -> Result<Vec<K8sBackupEntry>> {
    parse_backups(
        &kubectl(&["get", "backups.velero.io", "-A", "-o", "json"])?,
        "Velero",
    )
}

pub fn get_cluster_nodes() -> Result<Vec<ClusterNode>> {
    let parsed: Value = serde_json::from_str(&kubectl(&["get", "nodes", "-o", "json"])?)?;
    let items = parsed
        .get("items")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    Ok(items
        .into_iter()
        .map(|item| {
            let metadata = item.get("metadata").cloned().unwrap_or(Value::Null);
            let status = item.get("status").cloned().unwrap_or(Value::Null);
            let labels = metadata
                .get("labels")
                .and_then(Value::as_object)
                .cloned()
                .unwrap_or_default();
            let internal_ip =
                status
                    .get("addresses")
                    .and_then(Value::as_array)
                    .and_then(|addresses| {
                        addresses.iter().find_map(|address| {
                            (address.get("type").and_then(Value::as_str) == Some("InternalIP"))
                                .then(|| {
                                    address
                                        .get("address")
                                        .and_then(Value::as_str)
                                        .map(ToOwned::to_owned)
                                })
                                .flatten()
                        })
                    });
            let role = if labels.contains_key("node-role.kubernetes.io/control-plane")
                || labels.contains_key("node-role.kubernetes.io/master")
            {
                "controlplane".to_string()
            } else {
                "worker".to_string()
            };

            ClusterNode {
                name: metadata
                    .get("name")
                    .and_then(Value::as_str)
                    .unwrap_or("unknown")
                    .to_string(),
                internal_ip,
                role,
            }
        })
        .collect())
}

fn parse_backups(body: &str, source_type: &str) -> Result<Vec<K8sBackupEntry>> {
    let parsed: Value = serde_json::from_str(body)?;
    let items = parsed
        .get("items")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    Ok(items
        .into_iter()
        .map(|item| {
            let metadata = item.get("metadata").cloned().unwrap_or(Value::Null);
            let status = item.get("status").cloned().unwrap_or(Value::Null);
            let spec = item.get("spec").cloned().unwrap_or(Value::Null);
            K8sBackupEntry {
                name: metadata
                    .get("name")
                    .and_then(Value::as_str)
                    .unwrap_or("unknown")
                    .to_string(),
                namespace: metadata
                    .get("namespace")
                    .and_then(Value::as_str)
                    .unwrap_or("default")
                    .to_string(),
                source_type: source_type.to_string(),
                status: if source_type == "VolSync" {
                    parse_volsync_status(&status)
                } else {
                    status
                        .get("phase")
                        .and_then(Value::as_str)
                        .unwrap_or("unknown")
                        .to_string()
                },
                schedule: if source_type == "VolSync" {
                    spec.get("trigger")
                        .and_then(|value| value.get("schedule"))
                        .and_then(Value::as_str)
                        .unwrap_or("-")
                        .to_string()
                } else {
                    spec.get("scheduleName")
                        .and_then(Value::as_str)
                        .unwrap_or("-")
                        .to_string()
                },
                last_run: if source_type == "VolSync" {
                    status
                        .get("lastSyncTime")
                        .and_then(Value::as_str)
                        .unwrap_or("-")
                        .to_string()
                } else {
                    status
                        .get("completionTimestamp")
                        .and_then(Value::as_str)
                        .unwrap_or("-")
                        .to_string()
                },
            }
        })
        .collect())
}

fn parse_volsync_status(status: &Value) -> String {
    let conditions = status
        .get("conditions")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    for condition in conditions {
        if condition.get("type").and_then(Value::as_str) == Some("Synchronizing") {
            return if condition.get("status").and_then(Value::as_str) == Some("True") {
                "Syncing".to_string()
            } else {
                "Idle".to_string()
            };
        }
    }
    "unknown".to_string()
}

fn kubectl(extra: &[&str]) -> Result<String> {
    let kubectl = command::resolve_binary("kubectl", "PVT_KUBECTL_BIN")?;
    let mut argv = vec![kubectl];
    argv.extend(extra.iter().map(|value| value.to_string()));
    if let Some(kubeconfig) = discover_kubeconfig() {
        argv.push("--kubeconfig".to_string());
        argv.push(kubeconfig);
    }
    command::run(&argv, 512 * 1024)
}

pub(crate) fn discover_kubeconfig() -> Option<String> {
    if let Some(path) = env::var_os("KUBECONFIG") {
        let path = PathBuf::from(path);
        if path.is_file() {
            return Some(path.to_string_lossy().into_owned());
        }
    }

    let home = env::var_os("HOME")?;
    let candidate = PathBuf::from(home).join(".kube/config");
    if candidate.is_file() {
        return Some(candidate.to_string_lossy().into_owned());
    }
    None
}
